<launch>
  <arg name="main" default="true" />
  <arg name="json_dir" />
  <arg name="rviz" default="true" />
  <arg name="moveit" default="true" />
  <arg name="smach_viewer" default="false" />

  <!-- set bin_contents rosparam -->
  <node name="set_tote_contents"
        pkg="jsk_arc2017_common" type="set_location_in_rosparam.py">
    <remap from="~param" to="/tote_contents" />
    <rosparam subst_value="true">
      location: tote
      json_dir: $(arg json_dir)
    </rosparam>
  </node>

  <!-- state_server -->
  <node pkg="jsk_arc2017_baxter" type="state_server.py" name="state_server">
    <rosparam>
      is_pick: false
    </rosparam>
  </node>

  <!-- json_saver -->
  <node name="json_saver"
        pkg="jsk_arc2017_common" type="json_saver.py">
    <rosparam subst_value="true">
      json_dir: $(arg json_dir)
      output_dir: $(find jsk_arc2017_baxter)/output/stow_task
    </rosparam>
  </node>
  <node name="visualize_json"
        pkg="jsk_arc2017_common" type="visualize_json.py">
    <remap from="~input/json_dir" to="json_saver/output/json_dir" />
    <rosparam>
      types:
        - item_location
        - order
    </rosparam>
  </node>
  <node name="xdisplay_image_topic"
        pkg="jsk_baxter_startup" type="xdisplay_image_topic.py"
        args="/visualize_json/output/item_location_viz" />

  <!-- candidates publisher -->
  <group ns="left_hand_camera">
    <node name="candidates_publisher"
        pkg="jsk_arc2017_common" type="candidates_publisher.py" >
      <remap from="~input/json_dir" to="/json_saver/output/json_dir" />
      <remap from="~label_names" to="label_names" />
    </node>
  </group>
  <group ns="right_hand_camera">
    <node name="candidates_publisher"
        pkg="jsk_arc2017_common" type="candidates_publisher.py" >
      <remap from="~input/json_dir" to="/json_saver/output/json_dir" />
      <remap from="~label_names" to="label_names" />
    </node>
  </group>

  <!-- smach viewer -->
  <node pkg="smach_viewer" type="smach_viewer.py" name="smach_viewer" if="$(arg smach_viewer)"/>

  <!-- parameter -->
  <param name="/left_hand/target_bin" value="" />
  <param name="/right_hand/target_bin" value="" />

  <node name="rqt_yn_btn"
        pkg="jsk_rqt_plugins" type="rqt_yn_btn">
    <remap from="~" to="/rviz/yes_no_button" />
  </node>

  <group if="$(arg rviz)">
    <node name="rviz"
          pkg="rviz" type="rviz"
          args="-d $(find jsk_arc2017_baxter)/rvizconfig/stow.rviz">
    </node>
  </group>

  <!-- robot main -->
  <group if ="$(arg main)">
    <node unless="$(arg moveit)" pkg="jsk_arc2017_baxter" type="stow-main.l" name="left_hand_main"
          args="&quot;(progn (stow-init :ctype :larm-head-controller :moveit nil) (stow-mainloop :larm))&quot;"
          respawn="true" output="screen">
      <env name="DISPLAY" value="" />
    </node>

    <node unless="$(arg moveit)" pkg="jsk_arc2017_baxter" type="stow-main.l" name="right_hand_main"
          args="&quot;(progn (stow-init :ctype :rarm-head-controller :moveit nil) (stow-mainloop :rarm))&quot;"
          respawn="true" output="screen">
      <env name="DISPLAY" value="" />
    </node>

    <node if="$(arg moveit)" pkg="jsk_arc2017_baxter" type="stow-main.l" name="left_hand_main"
          args="&quot;(progn (stow-init :ctype :larm-head-controller :moveit t) (stow-mainloop :larm))&quot;"
          respawn="true" output="screen">
      <env name="DISPLAY" value="" />
    </node>

    <node if="$(arg moveit)" pkg="jsk_arc2017_baxter" type="stow-main.l" name="right_hand_main"
          args="&quot;(progn (stow-init :ctype :rarm-head-controller :moveit t) (stow-mainloop :rarm))&quot;"
          respawn="true" output="screen">
      <env name="DISPLAY" value="" />
    </node>
  </group>

  <!-- visualization -->
  <!-- TODO: json visualization -->
  <!--
  <node name="visualize_json"
    pkg="jsk_apc2016_common" type="visualize_stow_json.py"
    args="$(arg json)" />
  <node name="team_k_desktop_xdisplay"
        pkg="jsk_baxter_startup" type="xdisplay_image_topic.py"
        args="/visualize_json/output">
  </node>
  <node name="rosconsole_overlay_text"
        pkg="jsk_rviz_plugins" type="rosconsole_overlay_text.py">
    <rosparam>
      nodes:
        - /left_hand
        - /right_hand
    </rosparam>
  </node>
  -->

</launch>
